src/data/polynomial/denominators.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2020 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.polynomial.erase_lead
+/-!
+# Denominators of evaluation of polynomials at ratios
+
+Let `i : R → K` be a homomorphism of semirings.  Assume that `K` is commutative.  If `a` and
+`b` are elements of `R` such that `i b ∈ K` is invertible, then for any polynomial
+`f ∈ polynomial R` the "mathematical" expression `b ^ f.nat_degree * f (a / b) ∈ K` is in
+the image of the homomorphism `i`.
+-/
+
+open polynomial finset
+
+section denoms_clearable
+
+variables {R K : Type*} [semiring R] [comm_semiring K] {i : R →+* K}
+variables {a b : R} {bi : K}
+-- TODO: use hypothesis (ub : is_unit (i b)) to work with localizations.
+
+/-- `denoms_clearable` formalizes the property that `b ^ N * f (a / b)`
+does not have denominators, if the inequality `f.nat_degree ≤ N` holds.
+
+In the implementation, we also use provide an inverse in the existential.
+-/
+def denoms_clearable (a b : R) (N : ℕ) (f : polynomial R) (i : R →+* K) :=
+  ∃ D : R, ∃ bi : K, bi * i b = 1 ∧ i D = i b ^ N * eval (i a * bi) (polynomial.map i f)
+
+lemma denoms_clearable_zero (N : ℕ) (a b : R) (bu : bi * i b = 1) :
+  denoms_clearable a b N 0 i :=
+⟨0, bi, bu, by simp only [eval_zero, ring_hom.map_zero, mul_zero, map_zero]⟩
+
+lemma denoms_clearable_C_mul_X_pow (N : ℕ) (a b : R) (bu : bi * i b = 1) (n : ℕ) (r : R)
```

Comment:
```suggestion
lemma denoms_clearable_C_mul_X_pow {N : ℕ} (a : R) (bu : bi * i b = 1) {n : ℕ} (r : R)
```

Reply:
I accept these changes.  Why did you not suggest making `r` also implicit?

Reply:
On second thought, I would also like to see @jcommelin 's opinion: he has a much better idea of how this might be used in practice in Lean and he was the one who suggested which arguments to make implicit and which ones to make explicit.  Still, Bryan, I pushed your suggestion through!

Reply:
The general rule I go by is: if an argument can be inferred from a later one (e.g. it is a variable which appears in a later hypothesis), then it can be made implicit. There are two exceptions:
- Dot notation only works with explicit arguments, so that could be a reason to leave an argument explicit
- An argument appears on both sides of an `iff` then it can be made implicit as well

In this case, `r` cannot be inferred from the other arguments so I left it explicit.

By the way, if you click on the "Files changed" tab, you can add the suggestions to a batch and then just commit once (this is especially useful when some suggestions depend on others, so you don't get a bunch of broken builds in between).

Reply:
First of all, I did not know about the "batch commit" option: I will use it in the future!

As for the implicit arguments, I have a follow up question.  The variable `r` appears in the `denoms_clearable` argument as `(C r * X ^ n)`: is this too much inside a term to be inferred by Lean?

Finally, I believe that Johan's idea was the hypothesis `bu` would normally not be in context when using this function, so that giving `b` explicitly might have been necessary.  However, I still do not have enough practice with this to know how usage will play out (if it will!).

Reply:
`denoms_clearable` is after the colon and not before a `->`, so it is not a hypothesis.

Johan may very well be right! All I can say is that nothing that appeared so far had any issue with inferring `b`, but of course the theorems in this PR may not be representative of what's to come.

Reply:
My intuition is that when setting up the API the `bu` hypothesis will always be around, and so an implicit `b` is not much trouble. But when you want to apply the API, *I guess* that typically you will have `b` in your hands, and want `bu` to become a new goal.

Reply:
Ah, right. I didn't think about `apply`-ing. However, won't `b` always be inferrable from the goal?

Reply:
Well, maybe I didn't really mean `apply`. We'll have to see what happens in practice. But I wouldn't be surprised if relatively often we want to do `have : denoms_clearable ...`